File: relay/src/main/java/net/betrayd/webspeak/relay/ClientPlayerConnection.java
package net.betrayd.webspeak.relay;

import org.eclipse.jetty.websocket.api.Callback;
import org.eclipse.jetty.websocket.api.Session;
import org.eclipse.jetty.websocket.api.StatusCode;
import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose;
import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage;
import org.eclipse.jetty.websocket.api.annotations.OnWebSocketOpen;
import org.eclipse.jetty.websocket.api.annotations.WebSocket;

import net.betrayd.webspeak.relay.ServerConnection.LinkedConnection;

/**
 * The client-bound side of a player connection.
 */
@WebSocket
public class ClientPlayerConnection {
    private Session session;

    // store this here to handle a lot of packets fast without cpu doing map stuff.
    private ServerPlayerConnection serverConnection = null;
    private final String serverID;

    public ClientPlayerConnection(String server) {
        this.serverID = server;
    }

    public void send(String message) {
        session.sendText(message, Callback.NOOP);
    }

    public void close(int statusCode, String reason) {
        session.close(statusCode, reason, Callback.NOOP);
    }

    @OnWebSocketOpen
    public synchronized void onWebSocketOpen(Session session) {
        if (this.serverID == null) {
            session.close(StatusCode.BAD_DATA, "No server ID was supplied.", Callback.NOOP);
            return;
        }

        this.session = session;

        ServerConnection server = WebSpeakRelay.servers.get(serverID);
        if (server == null) {
            session.close(StatusCode.BAD_DATA,
                    "The server being connected to does not exist",Callback.NOOP);
            return;
        }
        String sessionID = NetUtils.splitQueryString(session.getUpgradeRequest().getQueryString()).get("id");
        LinkedConnection link = server.connections.get(sessionID);

        if (link == null) {
            session.close(StatusCode.BAD_PAYLOAD, "No player exists with session ID " + sessionID, Callback.NOOP);
        }
        serverConnection = link.serverConnection;
        serverConnection.connectedClient(session.getRemoteSocketAddress().toString(), this);
        link.clientConnection = this;
    }

    @OnWebSocketClose
    public synchronized void onWebSocketClose(int statusCode, String reason) {
        if (serverConnection == null) {
            return;
        }
        serverConnection.disconnectClient(statusCode, reason);
    }

    @OnWebSocketMessage
    public void onWebSocketMessage(Session session, String message) {
        this.serverConnection.send(message);
    }
}

File: clientPlayerConnection.ts
import { ServerPlayerConnection } from "./serverPlayerConnection.ts";
import { ServerConnection } from "./serverConnection.ts";
import { LinkedConnection } from "./linkedConnection.ts";
import { servers } from "./index.ts";

export class ClientPlayerConnection {

    socket!: WebSocket;
    serverID!: string | undefined;
    serverConnection: ServerPlayerConnection | null = null;
    origin!: string| null;

    constructor(socket: WebSocket, server: string | undefined, origin: string | null) {
        socket.addEventListener("open", this.eventOpen);
        socket.addEventListener("message", this.eventMessage);
        socket.addEventListener("close", this.eventClose);
        this.socket = socket;
        this.serverID = server;
        this.origin = origin;
    }

    public disconnect(code?: number, reason?: string) {
        this.socket.close(code, reason);
    }

    public send(message: string) {
        this.socket.send(message);
    }

    private eventOpen() {
        if (this.serverID == null) {
            this.socket.close(1002, "No server ID was supplied.");
            throw new Error("No server ID was supplied.");
        }

        const server: ServerConnection | undefined = servers.get(this.serverID);
        if (server == undefined) {
            this.socket.close(1002, "The server being connected to does not exist");
            throw new Error("The server being connected to does not exist");
        }

        const sessionID = new URL(this.socket.url).searchParams.get("id");
        if (sessionID == null) {
            this.socket.close(1002, "No ID provided");
            throw new Error("No ID provided");
        }
        const link: LinkedConnection | undefined = server.connections.get(sessionID);
        if (link == undefined) {
            this.socket.close(1002, "No player exists with session ID " + sessionID);
            throw new Error("No player exists with session ID " + sessionID);
        }
        this.serverConnection = link.serverConnection;
        if (this.origin == null) {
            this.socket.close(1002, "No origin");
            throw new Error("No origin");
        }
        this.serverConnection.connectClient(this.origin, this);
        link.clientConnection = this;
    }

    private eventMessage(event: MessageEvent<string>) {
        if (this.serverConnection == null) { return; }
        this.serverConnection.send(event.data);
    }

    private eventClose(event: CloseEvent) {
        if (this.serverConnection == null) { return; }
        this.serverConnection.disconnectClient(event.code, event.reason);
    }

}